libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp

```diff
@@ -0,0 +1,154 @@
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define ALLOW_DOUBLE_MATH_FUNCTIONS
+
+#include "AP_RangeFinder_config.h"
+
+#if AP_RANGEFINDER_AINSTEIN_LR_D1_ENABLED
+
+#include "AP_RangeFinder_Ainstein_LR_D1.h"
+
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_HAL/utility/sparse-endian.h>
+#include <GCS_MAVLink/GCS.h>
+
+static const uint8_t PACKET_HEADER_MSB = 0xEB;
+static const uint8_t PACKET_HEADER_LSB = 0x90;
+
+// make sure we know what size the packet object is:
+// assert_storage_size<AP_RangeFinder_Ainstein_LR_D1::LRD1Union::LRD1Packet, 32> _assert_storage_lrd1_packet;
+
+bool AP_RangeFinder_Ainstein_LR_D1::move_signature_in_buffer(uint8_t start)
+{
+    for (uint8_t i=start; i<buffer_used; i++) {
+        if (u.buffer[i] == PACKET_HEADER_MSB) {
+            memmove(&u.buffer[0], &u.buffer[i], buffer_used-i);
+            buffer_used -= i;
+            return true;
+        }
+    }
+    // header byte not in buffer
+    buffer_used = 0;
+    return false;
+}
+
+uint8_t AP_RangeFinder_Ainstein_LR_D1::LRD1Union::calculate_checksum() const
+{
+    uint32_t sum = 0;
+    for (uint8_t i=3; i<31; i++) {
+        sum += buffer[i];
+    }
+    return sum & 0xff;
+}
+
+// get_reading - read a value from the sensor
+bool AP_RangeFinder_Ainstein_LR_D1::get_reading(float &reading_m)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    const uint8_t num_read = uart->read(&u.buffer[buffer_used], ARRAY_SIZE(u.buffer)-buffer_used);
+    buffer_used += num_read;
+
+    if (buffer_used == 0) {
+        return false;
+    }
+
+    if (u.packet.header_msb != PACKET_HEADER_MSB &&
+        !move_signature_in_buffer(1)) {
+        return false;
+    }
+
+    if (buffer_used < sizeof(u.packet)) {
+        return false;
+    }
+
+    // sanity checks; see data sheet on these fixed values.
+    if (u.packet.header_lsb != PACKET_HEADER_LSB ||
+        u.packet.device_id != 0x00 ||
+        u.packet.length != 28 ||
+        u.calculate_checksum() != u.packet.checksum) {
+        // sanity checks failed - discard and try again next time we're called:
+        move_signature_in_buffer(1);
+        return false;
+    }
+
+    reading_m = be16toh(u.packet.object1_alt) * 0.01;
+
+    // consume this packet:
+    move_signature_in_buffer(sizeof(u.packet));
+
+    /*  From datasheet:
+      Altitude data from radar may have unexpected
+      or incorrect readings if aircraft pitch and/or
+      roll are beyond the radar’s detection angle
+      (Azimuth 43°，Elevation 30°). The ideal case is
+      keeping radar perpendicular to ground.
+    */
+#ifndef HAL_BUILD_AP_PERIPH   // we may need to move this elsewhere to the entire function or driver?
+    const auto &ahrs = AP::ahrs();
+    if (fabs(ahrs.roll) > radians(43)) {
+        return false;
+    }
+    if (fabs(ahrs.pitch) > radians(30)) {
+        return false;
+    }
+#endif
```

Comment:
This should be AP_AHRS_ENABLED instead of a Periph check but really from a driver's perspective this chunk should be removed. There's no guarantee this is used to look straight down at the ground, it could be forward/side/up or whatever orientation. That can be checked via `enum Rotation` The datasheet just says that for proper operation it expects to bounce off a roughly perpendicular surface.. but that's not the driver's responsibility to enforce.